lib/charms/dns_transfer/v0/dns_transfer.py

```diff
@@ -0,0 +1,346 @@
+
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "908bcd1f0ad14cabbc9dca25fa0fc87c"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 5
+
+PYDEPS = ["pydantic>=2"]
+
+# pylint: disable=wrong-import-position
+import json
+import logging
+import re
+from enum import Enum
+from typing import Any, Optional, cast
+
+import ops
+import pydantic
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "dns-transfer"
+VALID_NAME_RFC952 = re.compile(r"^[a-zA-Z]([a-zA-Z0-9-]+[.]?)*[a-zA-Z0-9]$")
+
+
+class TransportSecurity(str, Enum):
+    """Represent the transport security values.
+
+    Attributes:
+        TCP: tcp
+        TLS: tls
+    """
+
+    TCP = "tcp"
+    TLS = "tls"
+
+
+class DNSTransferProviderData(pydantic.BaseModel):
+    """List of information for the provider to manage and transfer.
+
+    Attributes:
+        addresses: IP list of the units composing the provider's deployment.
+        transport: Type of transport (tls or tcp).
+        remote_hostname: Advertised name of the host in the TLS certificate (optional).
+        zones: List of zone names.
+    """
+
+    addresses: list[pydantic.IPvAnyAddress]
+    transport: TransportSecurity
+    remote_hostname: Optional[str] = None
+    zones: list[str]
+
+    # pydantic wants 'self' as first argument
+    @pydantic.field_validator("addresses", mode="before")
+    def deduplicate_addresses(cls, v: Any) -> list:  # noqa: N805 pylint: disable=E0213
+        """Deduplicate addresses.
+
+        Args:
+            v: value
+
+        Returns:
+            list with unique values.
+        """
+        return list(set(v))
+
+    # pydantic wants 'self' as first argument
+    @pydantic.field_validator("zones", mode="before")
+    def deduplicate_zones(cls, v: Any) -> list:  # noqa: N805 pylint: disable=E0213
+        """Deduplicate zones.
+
+        Args:
+            v: value
+
+        Returns:
+            list with unique values.
+        """
+        return list(set(v))
+
+    # pydantic wants 'self' as first argument
+    @pydantic.field_validator("remote_hostname")
+    def validate_hostname(cls, v: Any) -> Any:  # noqa: N805 pylint: disable=E0213
+        """Deduplicate zones.
+
+        Args:
+            v: The input value provided for the `remote_hostname` field.
+
+        Returns:
+            A validated hostname.
+
+        Raises:
+            ValueError: if hostname is not RFC 952 valid name.
+        """
+        if v is None:
+            return v
+        if len(v) > 24:
+            raise ValueError("remote_hostname exceeds 24 characters (RFC 952)")
+        if not VALID_NAME_RFC952.fullmatch(v):
+            raise ValueError("remote_hostname does not match RFC 952 rules")
+        return v
```

Comment:
I was looking at this and discussed it with Sebastien a bit. Actually, RFC952 is obsoleted by RFC1123 which is obsoleted again by another and so on. In the end, you actually can use the same validator as the one in `dns_authority`: https://github.com/canonical/dns-operators/blob/dns-resolver-lib/lib/charms/dns_authority/v0/dns_authority.py#L64

Reply:
Maybe we could start the library of validators right now

Reply:
I agree, we could work on that in an independent story/PR.